### PR TITLE
Implement Reflect for PhantomData

### LINKED ISSUE #5144: PhantomData does not implement Reflect
## What problem does this solve or what need does it fill?

PhantomData is a common and useful type, it should implement the Reflect trait.

Users cannot do this themselves because of orphan rules, and wrappers are very frustrating here.

## What solution would you like?

Implement Reflect for PhantomData in bevy_reflect.

### PR DESCRIPTION
**Pull request by BenjaminBrienen**
# Objective

- Fixed #5144

## Solution

- impl_reflect! following the example

## Testing

N/A
I don't see tests for the other `core` `impl_reflect`s.

### CODE REVIEW

File: crates/bevy_reflect/src/impls/std.rs [hunk 1]
```diff
@@ -1484,6 +1485,11 @@ impl_reflect! {
     }
 }
 
+impl_reflect! {
```
Comment (MrGVSV) [inline crates/bevy_reflect/src/impls/std.rs:1488]: Any reason we're not using `impl_reflect_value!` here? I feel like this makes more sense as a `Value` type, no?

Comment (BenjaminBrienen) [reply crates/bevy_reflect/src/impls/std.rs:1488]: I got an error when doing that. I can take another stab at it.

**Comment by MrGVSV** [comment]
> [#5145 (comment)](https://github.com/bevyengine/bevy/pull/5145#issuecomment-1327108931)
> 
> I'm curious why we want this now and not before.

It's still the same issues as before. I think the biggest problem is serialization. It really only adds noise to the serialized output. And if we make it `Reflect`, then it becomes easy for library authors to forget to ignore it themselves, which means downstream users will see it if they serialize it.

If do this, we may want to consider skipping it during serialization (and other operations such as `reflect_partial_eq`).

File: crates/bevy_reflect/src/impls/std.rs (hunk 1 above)
Comment (MrGVSV) [reply crates/bevy_reflect/src/impls/std.rs:1488]: What was the error?

File: crates/bevy_reflect/src/impls/std.rs
Comment (MrGVSV) [inline crates/bevy_reflect/src/impls/std.rs:1]: Also, can we add a serialization test for `PhantomData` in the `serde` module? Whether or not we're allowing this type to be serialized, we should ensure any type containing it can be deserialized and `FromReflect`-ed.

**Review by MrGVSV — CHANGES_REQUESTED** [review]

File: crates/bevy_reflect/src/impls/std.rs (hunk 1 above)
Comment (BenjaminBrienen) [reply crates/bevy_reflect/src/impls/std.rs]: It was something about PhantomData not being able to provide compilation information. I don't get it now though.

**Comment by soqb** [comment]
do we want really need this? i feel like `PhantomData` is supposed to act as if its not there at all (its a :ghost: after all). is there anything not accomplished by annotating `#[reflect(ignore)]` or is it simply an ergonomic improvement? i'd really rather not have all the "noise" that `@MrGVSV` mentioned (not just in serialization mind, also in runtime and binary size) just for the sake of a few less tokens to type out.

tldr; the implementation is correct but i'm not sure the feature is desirable.

**Comment by BenjaminBrienen** [comment]
Maybe it would be better to instead implement a way to always ignore marker types?

**Comment by BenjaminBrienen** [comment]
I don't know how to fix the unit test. Someone else can give me a pointer or do it themselves if they want. :)

**Comment by hymm** [comment]
Could there be a way of saying to ignore a whole type when you register it?

**Comment by alice-i-cecile** [comment]
Yeah, that would be my preferred solution here.
